### PR TITLE
 DCOS_OSS-5406: (BETA testing) windows-instance module, Azure provider.

### PR DESCRIPTION
**Pull request by sergiimatus**
@sebbrandt87 ,
please review/comment/approve. 
That's core Windows-Instance module, which was splitted from Instance module

### CODE REVIEW

File: winrm_setup.ps1 [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+# Configure WinRM on Azure Windows instance
+& $winrm_cmd
+
+$hostname = $(Invoke-RestMethod -Headers @{"Metadata"="true"} -URI "http://169.254.169.254/metadata/instance/compute/name?api-version=2019-02-01&&format=text")
+
+# Generate self signed certificate for WinRM usage
+New-SelfSignedCertificate -DnsName $hostname -CertStoreLocation Cert:\LocalMachine\My
+New-Item WSMan:\localhost\Listener -Address * -Transport HTTPS -HostName $hostname -CertificateThumbPrint $(ls Cert:\LocalMachine\My).Thumbprint -Force
+Set-NetFirewallProfile -Profile Domain,Public,Private -Enabled False
+Set-MpPreference -DisableRealtimeMonitoring $true
```
Comment (sebbrandt87) [inline winrm_setup.ps1]: missing linebreak

File: variables.tf [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  description = "User data to be used on these instances (cloud-init)"
+  default     = ""
+}
+
+variable "admin_username" {
+  description = "Windows admin user"
+  default     = "dcosadmin"
+}
+
+variable "admin_password" {
```
Comment (sebbrandt87) [inline variables.tf]: we do not set the admin password anywhere specific but generate it

File: variables.tf [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  description = "Windows admin password"
+  default     = []
+}
+
+variable "public_ssh_key" {
+  description = "SSH Public Key"
+}
+
+variable "allow_stopping_for_update" {
```
Comment (sebbrandt87) [inline variables.tf]: not used, use the same defaults for all non windows related values as in https://github.com/dcos-terraform/terraform-azurerm-instance/blob/support/0.2.x/variables.tf

File: main.tf [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    admin_password = "${element(random_password.password.*.result, count.index)}"
+    custom_data    = "${var.custom_data}"
+  }
+
+  os_profile_windows_config {
+    provision_vm_agent        = true
+    enable_automatic_upgrades = false
+
+    winrm = {
+      protocol = "http"
```
Comment (sebbrandt87) [inline main.tf]: should be https

File: variables.tf [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+variable "name_prefix" {
+  description = "Name Prefix"
+  default     = ""
+}
+
+variable "vm_size" {
+  description = "Azure virtual machine size"
+}
+
+variable "dcos_instance_os" {
```
Comment (sebbrandt87) [inline variables.tf:22]: default should be defined here, as it is a specific windows module

**Review by sebbrandt87 — CHANGES_REQUESTED** [review]

File: variables.tf (hunk 2 above)
Comment (sergiimatus) [reply variables.tf]: I was thinking to let user specify admin_password, however thought that it won't be secure. Removing in next commit

File: main.tf (hunk 4 above)
Comment (sergiimatus) [reply main.tf]: agreed, but https didn't work as I expected. Therefore used http. 
There is good example which supports https starting from terraform 0.12 unfortunately - https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/examples/virtual-machines/provisioners/windows/main.tf

File: variables.tf (hunk 5 above)
Comment (sergiimatus) [reply variables.tf:22]: agreed

File: variables.tf (hunk 3 above)
Comment (sergiimatus) [reply variables.tf]: agreed, removing

**Review by sergiimatus — COMMENTED** [review]
@sebbrandt87, 
- fixed WinRM (purely moved the setup to separate step -> extension using the PS1 script)
- fixed all other minor concerns

File: variables.tf (hunk 2 above)
Comment (sergiimatus) [reply variables.tf]: fixed

File: main.tf (hunk 4 above)
Comment (sergiimatus) [reply main.tf]: fixed by removing the block - setup will happen in extension

**Comment by sergiimatus** [comment]
@sebbrandt87, 
- fixed WinRM (purely moved the setup to separate step -> extension using the PS1 script)
- fixed all other minor concerns

File: main.tf [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+  virtual_machine_name = "${format(var.hostname_format, count.index + 1, local.cluster_name)}"
+  publisher            = "Microsoft.Compute"
+  type                 = "CustomScriptExtension"
+  type_handler_version = "1.8"
+  count                = "${var.num}"
+  depends_on           = ["azurerm_virtual_machine.windows_instance"]
+
+  settings = <<SETTINGS
+    {
+        "fileUris": ["https://raw.githubusercontent.com/dcos-terraform/terraform-azurerm-windows-instance/master/winrm_setup.ps1"],
```
Comment (sebbrandt87) [inline main.tf:176]: please use the `file` operation here for `commandToExecute` as we could not bet on github being available and want exactly the version of the script we package in each specific terraform module release.

**Review by sebbrandt87 — CHANGES_REQUESTED** [review]

**Comment by sebbrandt87** [comment]
please create a new PR against `support/0.2.x` branch

File: main.tf (hunk 6 above)
Comment (sergiimatus) [reply main.tf:176]: Tried `"fileUris": ["file(${path.module}/winrm_script.ps1)"]` and it doesn't work . 
Reason: as per [Microsoft documentation](https://docs.microsoft.com/en-us/azure/virtual-machines/extensions/custom-script-windows) - "Scripts can be downloaded from Azure storage or GitHub, or provided to the Azure portal at extension run time." Meaning, solution here is either to place content of winrm script using `cmd` syntax and execute line by line, or use as is - via github link, as shown in the example

Comment (sebbrandt87) [reply main.tf:176]: see https://github.com/dcos-terraform/terraform-azurerm-windows-instance/pull/2#discussion_r316873122

**Pull request closed without merging**